src/nativeops.jl

```diff
@@ -0,0 +1,336 @@
+#
+# Forward and backward functions call require more care because they can happen async, we
+# have to handle them in the task that wait for the functions to be called.
+#
+# Todo: Cleanup tasks.
+###
+immutable NDArrayOpInfo
+  forward :: Ptr{Void}
+  backward :: Ptr{Void}
+  infer_shape :: Ptr{Void}
+  list_outputs :: Ptr{Void}
+  list_arguments :: Ptr{Void}
+  declare_backward_dependecy :: Ptr{Void}
+
+  p_forward :: Ptr{Void}
+  p_backward :: Ptr{Void}
+  p_infer_shape :: Ptr{Void}
+  p_list_outputs :: Ptr{Void}
+  p_list_arguments :: Ptr{Void}
+  p_declare_backward_dependency :: Ptr{Void}
+
+  function NDArrayOpInfo(op :: Operator)
+    # infer_shape, list_args, list_outputs are called directly and use dynamic dispatch,
+    # for finding the correct operator.
+    p_is = p_lo = p_la = p_dbd = pointer_from_objref(op)
+
+    c_wrapper_fb = cfunction(_wrapper_fb, Bool, (Cint, Ptr{Ptr{Void}}, Ptr{Cint}, Ptr{Void}))
+    c_wrapper_infer = cfunction(_wrapper_infer, Bool, (Cint, Ptr{Cint}, Ptr{Ptr{Cuint}}, Ptr{Void}))
+    c_wrapper_list_outputs = cfunction(_wrapper_list_outputs, Bool, (Ptr{Ptr{Ptr{Cchar}}}, Ptr{Void}))
+    c_wrapper_list_arguments = cfunction(_wrapper_list_arguments, Bool, (Ptr{Ptr{Ptr{Cchar}}}, Ptr{Void}))
+    c_wrapper_declare_backward_dependency = cfunction(_wrapper_declare_backward_dependency, Bool, (Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Ptr{Cint}}, Ptr{Void}))
+
+    # Setting up for handling backward/forward. Each function has a condition for a task
+    # to wait on and a libuv callback that notifies that condition, to handle the call.
+    cond_forward = Condition()
+    cond_backward = Condition()
+    cb_f = Base.SingleAsyncWork(data -> notify(cond_forward))
+    cb_b = Base.SingleAsyncWork(data -> notify(cond_backward))
+
+    # The return value of the callbacks is stored in _FB and the first element is the
+    # libuv handle.
+    r_forward = Ref(_FB(cb_f.handle))
+    r_backward = Ref(_FB(cb_f.handle))
+
+    p_f = Base.unsafe_convert(Ptr{Void}, r_forward)
+    p_b = Base.unsafe_convert(Ptr{Void}, r_backward)
+
+    # Task for handling forward
+    @schedule begin
+      try
+        while true
+           wait(cond_forward)
+           cond_forward = Condition()
+           _entry_forward(op, r_forward[])
+        end
+      catch
+        rethrow()
+      finally
+        Base.close(cb_f)
+      end
+    end
+
+    # Task for handling backward
+    @schedule begin
+      try
+        while true
+           wait(cond_backward)
+           cond_backward = Condition()
+           _entry_backward(op, r_backward[])
+        end
+      catch
+        rethrow()
+      finally
+        Base.close(cb_f)
+      end
+    end
+
+    new(c_wrapper_fb, c_wrapper_fb, c_wrapper_infer, c_wrapper_list_outputs,
+        c_wrapper_list_arguments,  c_wrapper_declare_backward_dependency,
+        p_f, p_b, p_is, p_lo, p_la, p_dbd)
+  end
+end
+
+###
+# Infer and list are called in sync.
```

Comment:
Why those callbacks could be called in sync? Could functions like `unsafe_load` possibly trigger gc?